[SDK-3028] Add advanced used cases for Swift (#5057)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/swift/advanced_use_cases/locations_and_geofences.md

---

> nav_title: Locations and Geofences

Reviewer: General question: How does this section relate to turning on automatic location tracking (ala https://braze-docs-qlxwa9y1a-braze.vercel.app/docs/developer_guide/platform_integration_guides/swift/analytics/location_tracking/#enabling-automatic-location-tracking)

configuration.location.automaticLocationCollection = true
configuration.location.geofencesEnabled = true
configuration.location.automaticGeofenceRequests = true
let braze = Braze(configuration: configuration)
AppDelegate.braze = braze
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
BRZConfiguration *configuration =
    [[BRZConfiguration alloc] initWithApiKey:brazeApiKey
                                    endpoint:brazeEndpoint];
configuration.logger.level = BRZLoggerLevelInfo;
configuration.location.brazeLocationProvider = [[BrazeLocationProvider alloc] init];
configuration.location.automaticLocationCollection = YES;
configuration.location.geofencesEnabled = YES;
configuration.location.automaticGeofenceRequests = YES;
Braze *braze = [[Braze alloc] initWithConfiguration:configuration];
AppDelegate.braze = braze;
```

{% endtab %}
{% endtabs %}

## Step 4: Check for Braze background push

Braze syncs geofences to devices using background push notifications. Follow the [ignoring silent push][7] article to ensure that your application does not take any unwanted actions upon receiving Braze's geofence sync notifications.

> ## Step 5: Add NSLocationAlwaysUsageDescription to your Info.plist

Reviewer 2: ```suggestion
## Step 5: Add location usage description strings to your Info.plist
```

> Add the key `NSLocationAlwaysUsageDescription` and `NSLocationAlwaysAndWhenInUseUsageDescription` or `NSLocationWhenInUseUsageDescription` to your `info.plist` with a `String` value that has a description of why your application needs to track location.

Reviewer 2: ```suggestion
Add the key `NSLocationAlwaysUsageDescription`, `NSLocationAlwaysAndWhenInUseUsageDescription` or `NSLocationWhenInUseUsageDescription` to your `info.plist` with a `String` value that has a description of why your application needs to track location.
```

This description will be shown when the system location prompt requests authorization and should clearly explain the benefits of location tracking to your users.

## Step 6: Request authorization from the user

The geofences feature is only functional while `Always` location authorization or `AuthorizedWhenInUse` with the `Background Mode -> Location updates` capability enabled is granted.

To request for `Always` or `AuthorizedWhenInUse` location authorization, use the following code:

{% tabs %}
{% tab swift %}

```swift
var locationManager = CLLocationManager()

> locationManager.requestWhenInUseAuthorization()

Reviewer 2: ```suggestion
locationManager.requestWhenInUseAuthorization()
// or
```

locationManager.requestAlwaysAuthorization()
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
CLLocationManager *locationManager = [[CLLocationManager alloc] init];

> [locationManager requestWhenInUseAuthorization];
> [locationManager requestAlwaysAuthorization];

Reviewer 2: ```suggestion
[locationManager requestWhenInUseAuthorization];
// or
[locationManager requestAlwaysAuthorization];
```

Tool:
  - Location

---

# Locations and geofences for iOS

At the core of Braze’s real-time location offering is the concept of a [geofence]({{site.baseurl}}/user_guide/engagement_tools/locations_and_geofences#about-locations-and-geofences). A geofence is a virtual geographic area, represented as latitude and longitude combined with a radius, forming a circle around a specific global position.

This article covers setting up geofences for your iOS SDK integration. Geofences are only available in select Braze packages. Reach out to your Braze customer success manager to get started.

{% alert important %}
As of iOS 14, geofences do not work reliably for users who choose to give their approximate location permission.
{% endalert %}

## Step 1: Enable background push

To fully utilize our geofence syncing strategy, you must have [background push][6] enabled in addition to completing the standard push integration.

## Step 2: Enable Braze location services
Braze location services [must be enabled][1] through the SDK. They are not enabled by default.

## Step 3: Enable geofences

> Enable geofences by setting `location.geofencesEnabled` to `true` on the `configuration` object that initializes the[`Braze`][1] instance.

Reviewer 3: There's a couple of other `location` configurations mentioned in the code snippets. Should we provide a link to https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/braze/configuration-swift.class/location-swift.class to describe the rest of the configurations?

Author: We could; most of that should be covered in the locations article, but I can add a link

## Step 2: Enable geofences

Enable geofences by setting `location.geofencesEnabled` to `true` on the `configuration` object that initializes the[`Braze`][1] instance.
{% tabs %}
{% tab swift %}

```swift
let configuration = Braze.Configuration(

>   apiKey: "<BRAZE_API_KEY>",

Reviewer 3: I think we want to use `{}` instead of `<>` - @josh-mccrowell-braze for sanity check?

Reviewer: Yep, good callout! As a shorthand for "This is placeholder text to be replaced," {} is the tech writing team's shorthand for that!

## Disabling automatic geofence requests

You can disable automatic geofence requests in your `configuration` object passed to [`init(configuration)`][4]. Set `automaticGeofenceRequests` to `false`. For example:

{% tabs %}
{% tab swift %}

```swift
let configuration = Braze.Configuration(
  apiKey: "<BRAZE_API_KEY>",
  endpoint: "<BRAZE_ENDPOINT>"
)
configuration.automaticGeofencesRequest = false
let braze = Braze(configuration: configuration)
AppDelegate.braze = braze
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
BRZConfiguration *configuration =
  [[BRZConfiguration alloc] initWithApiKey:brazeApiKey
                                  endpoint:brazeEndpoint];

> configuration.automaticGeofencesRequest = false;

Reviewer 4: ```suggestion
configuration.automaticGeofencesRequest = NO;
```

> To control the location that the SDK reports for the purposes of receiving the most relevant geofences, starting in iOS SDK version 3.21.3, you can manually request geofences by providing the latitude and longitude of a location. It is recommended to disable automatic geofence requests when using this method. To do so, use the following code:

Reviewer: @nigelfaustino Is this reference to the iOS SDK version intentional? Could you please doublecheck the introduction to this section to make sure these recommendations are still accurate and up to date?

Author: I'll remove the version, that's unintentional
